// lazada-sdk/lib/LazadaRequest/index.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

/**
 * Determind if response is successful
 * @param {LazadaOpenPlatformAPIResponse} response
 * @return {boolean} success
 */


var _nodeFetch = require('node-fetch');

var _nodeFetch2 = _interopRequireDefault(_nodeFetch);

var _constants = require('./constants');

var _signature = require('./signature');

var _Common = require('../types/Common');

var _Request = require('./types/Request');

var _Response = require('./types/Response');

var _flowRuntime = require('flow-runtime');

var _flowRuntime2 = _interopRequireDefault(_flowRuntime);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LazadaOpenPlatformAPIResponse = _flowRuntime2.default.tdz(function () {
  return _Response.LazadaOpenPlatformAPIResponse;
});

var SDKRequestMetaData = _flowRuntime2.default.tdz(function () {
  return _Request.SDKRequestMetaData;
});

var SystemQueryParams = _flowRuntime2.default.tdz(function () {
  return _Request.SystemQueryParams;
});

var KeyValueDictionary = _flowRuntime2.default.tdz(function () {
  return _Common.KeyValueDictionary;
});

var isResponseSuccessful = _flowRuntime2.default.annotate(function isResponseSuccessful(response) {
  var _responseType = _flowRuntime2.default.union(_flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse), _flowRuntime2.default.any());

  var _returnType = _flowRuntime2.default.return(_flowRuntime2.default.boolean());

  _flowRuntime2.default.param('response', _responseType).assert(response);

  return _returnType.assert((typeof response === 'undefined' ? 'undefined' : _typeof(response)) === 'object' && response !== null && response.hasOwnProperty('code') && response.code === _constants.RESPONSE.SUCCESS.CODE);
}, _flowRuntime2.default.function(_flowRuntime2.default.param('response', _flowRuntime2.default.union(_flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse), _flowRuntime2.default.any())), _flowRuntime2.default.return(_flowRuntime2.default.boolean())));

var handleLazadaResponse = _flowRuntime2.default.annotate(function handleLazadaResponse(response, meta) {
  var _responseType2 = _flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse);

  var _metaType = _flowRuntime2.default.ref(SDKRequestMetaData);

  var _returnType2 = _flowRuntime2.default.return(_flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse));

  _flowRuntime2.default.param('response', _responseType2).assert(response);

  _flowRuntime2.default.param('meta', _metaType).assert(meta);

  // for debug only
  var _log_request = _flowRuntime2.default.annotate(function _log_request(_arg, response) {
    var _t$ref$assert = _flowRuntime2.default.ref(SDKRequestMetaData).assert(_arg),
        method = _t$ref$assert.method,
        apiPath = _t$ref$assert.apiPath,
        payload = _t$ref$assert.payload,
        query = _t$ref$assert.query;

    console.info('[%s] '.replace(/%s/, method) + apiPath, ' ', payload, ' ', query);
    console.log(JSON.stringify(response, null, 2));
  }, _flowRuntime2.default.function(_flowRuntime2.default.param('_arg', _flowRuntime2.default.ref(SDKRequestMetaData)), _flowRuntime2.default.param('response', _flowRuntime2.default.any())));
  // _log_request(meta, response)
  if (isResponseSuccessful(response)) {
    return Promise.resolve(response).then(function (_arg2) {
      return _returnType2.assert(_arg2);
    });
  } else {
    return Promise.reject(response).then(function (_arg3) {
      return _returnType2.assert(_arg3);
    });
  }
}, _flowRuntime2.default.function(_flowRuntime2.default.param('response', _flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse)), _flowRuntime2.default.param('meta', _flowRuntime2.default.ref(SDKRequestMetaData)), _flowRuntime2.default.return(_flowRuntime2.default.ref('Promise', _flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse)))));

var get = _flowRuntime2.default.annotate(function get(base, appKey, appSecret, apiPath, accessToken, params) {
  var _baseType = _flowRuntime2.default.string();

  var _appKeyType = _flowRuntime2.default.string();

  var _appSecretType = _flowRuntime2.default.string();

  var _apiPathType = _flowRuntime2.default.string();

  var _accessTokenType = _flowRuntime2.default.nullable(_flowRuntime2.default.string());

  var _paramsType = _flowRuntime2.default.ref(KeyValueDictionary);

  var _returnType3 = _flowRuntime2.default.return(_flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse));

  _flowRuntime2.default.param('base', _baseType).assert(base);

  _flowRuntime2.default.param('appKey', _appKeyType).assert(appKey);

  _flowRuntime2.default.param('appSecret', _appSecretType).assert(appSecret);

  _flowRuntime2.default.param('apiPath', _apiPathType).assert(apiPath);

  _flowRuntime2.default.param('accessToken', _accessTokenType).assert(accessToken);

  _flowRuntime2.default.param('params', _paramsType, true).assert(params);

  var qs = Object.assign({}, params, getSystemQueryParamObject(appKey, appSecret, apiPath, accessToken, params));
  var url = new URL(base + apiPath);
  Object.keys(qs).forEach(function (key) {
    return url.searchParams.append(key, params[key]);
  });

  return (0, _nodeFetch2.default)(url, {
    method: 'GET',
    qs: qs
  }).then(function (res) {
    return res.json();
  }).then(function (res) {
    return handleLazadaResponse(response, meta);
  }).then(function (_arg4) {
    return _returnType3.assert(_arg4);
  });
}, _flowRuntime2.default.function(_flowRuntime2.default.param('base', _flowRuntime2.default.string()), _flowRuntime2.default.param('appKey', _flowRuntime2.default.string()), _flowRuntime2.default.param('appSecret', _flowRuntime2.default.string()), _flowRuntime2.default.param('apiPath', _flowRuntime2.default.string()), _flowRuntime2.default.param('accessToken', _flowRuntime2.default.nullable(_flowRuntime2.default.string())), _flowRuntime2.default.param('params', _flowRuntime2.default.ref(KeyValueDictionary)), _flowRuntime2.default.return(_flowRuntime2.default.ref('Promise', _flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse)))));

var post = _flowRuntime2.default.annotate(function post(base, appKey, appSecret, apiPath, accessToken, body) {
  var _baseType2 = _flowRuntime2.default.string();

  var _appKeyType2 = _flowRuntime2.default.string();

  var _appSecretType2 = _flowRuntime2.default.string();

  var _apiPathType2 = _flowRuntime2.default.string();

  var _accessTokenType2 = _flowRuntime2.default.nullable(_flowRuntime2.default.string());

  var _bodyType = _flowRuntime2.default.ref(KeyValueDictionary);

  var _returnType4 = _flowRuntime2.default.return(_flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse));

  _flowRuntime2.default.param('base', _baseType2).assert(base);

  _flowRuntime2.default.param('appKey', _appKeyType2).assert(appKey);

  _flowRuntime2.default.param('appSecret', _appSecretType2).assert(appSecret);

  _flowRuntime2.default.param('apiPath', _apiPathType2).assert(apiPath);

  _flowRuntime2.default.param('accessToken', _accessTokenType2).assert(accessToken);

  _flowRuntime2.default.param('body', _bodyType, true).assert(body);

  // turns out even it's HTTP POST, Lazada expect `body` to be part of query string
  var qs = Object.assign({}, body, getSystemQueryParamObject(appKey, appSecret, apiPath, accessToken, body));
  var url = new URL(base + apiPath);
  Object.keys(qs).forEach(function (key) {
    return url.searchParams.append(key, params[key]);
  });

  return (0, _nodeFetch2.default)(url, {
    method: 'GET',
    body: body
  }).then(function (res) {
    return res.json();
  }).then(function (res) {
    return handleLazadaResponse(response, meta);
  }).then(function (_arg5) {
    return _returnType4.assert(_arg5);
  });
}, _flowRuntime2.default.function(_flowRuntime2.default.param('base', _flowRuntime2.default.string()), _flowRuntime2.default.param('appKey', _flowRuntime2.default.string()), _flowRuntime2.default.param('appSecret', _flowRuntime2.default.string()), _flowRuntime2.default.param('apiPath', _flowRuntime2.default.string()), _flowRuntime2.default.param('accessToken', _flowRuntime2.default.nullable(_flowRuntime2.default.string())), _flowRuntime2.default.param('body', _flowRuntime2.default.ref(KeyValueDictionary)), _flowRuntime2.default.return(_flowRuntime2.default.ref('Promise', _flowRuntime2.default.ref(LazadaOpenPlatformAPIResponse)))));

/**
 * Gather system and business parameters to compute signature
 * @param {string} appKey
 * @param {string} appSecret
 * @param {string} apiPath
 * @param {string?} accessToken
 * @param {KeyValueDictionary?} payload
 * @return {SystemQueryParams}
 */
var getSystemQueryParamObject = _flowRuntime2.default.annotate(function getSystemQueryParamObject(appKey, appSecret, apiPath, accessToken, payload) {
  var _appKeyType3 = _flowRuntime2.default.string();

  var _appSecretType3 = _flowRuntime2.default.string();

  var _apiPathType3 = _flowRuntime2.default.string();

  var _accessTokenType3 = _flowRuntime2.default.nullable(_flowRuntime2.default.string());

  var _payloadType = _flowRuntime2.default.ref(KeyValueDictionary);

  var _returnType5 = _flowRuntime2.default.return(_flowRuntime2.default.ref(SystemQueryParams));

  _flowRuntime2.default.param('appKey', _appKeyType3).assert(appKey);

  _flowRuntime2.default.param('appSecret', _appSecretType3).assert(appSecret);

  _flowRuntime2.default.param('apiPath', _apiPathType3).assert(apiPath);

  _flowRuntime2.default.param('accessToken', _accessTokenType3).assert(accessToken);

  _flowRuntime2.default.param('payload', _payloadType, true).assert(payload);

  var systemParams = _flowRuntime2.default.object(_flowRuntime2.default.property('app_key', _flowRuntime2.default.string()), _flowRuntime2.default.property('timestamp', _flowRuntime2.default.string()), _flowRuntime2.default.property('sign_method', _flowRuntime2.default.string()), _flowRuntime2.default.property('access_token', _flowRuntime2.default.string(), true)).assert({
    app_key: appKey,
    timestamp: Date.now().toString(),
    sign_method: 'sha256'
  });

  if (accessToken) {
    systemParams.access_token = accessToken;
  }

  return _returnType5.assert(Object.assign({}, {
    sign: (0, _signature.signRequest)(appSecret, apiPath, Object.assign({}, payload, systemParams))
  }, systemParams));
}, _flowRuntime2.default.function(_flowRuntime2.default.param('appKey', _flowRuntime2.default.string()), _flowRuntime2.default.param('appSecret', _flowRuntime2.default.string()), _flowRuntime2.default.param('apiPath', _flowRuntime2.default.string()), _flowRuntime2.default.param('accessToken', _flowRuntime2.default.nullable(_flowRuntime2.default.string())), _flowRuntime2.default.param('payload', _flowRuntime2.default.ref(KeyValueDictionary)), _flowRuntime2.default.return(_flowRuntime2.default.ref(SystemQueryParams))));

// constructor
module.exports = {
  get: get,
  post: post,
  isResponseSuccessful: isResponseSuccessful,
  // helper
  getSystemQueryParamObject: getSystemQueryParamObject
};
